protocols/stream/src/behaviour.rs

```diff
@@ -0,0 +1,287 @@
+};
+
+use crate::{
+    handler::{Handler, NewStream, RegisterProtocol},
+    Control, IncomingStreams,
+};
+
+/// A generic behaviour for stream-oriented protocols.
+pub struct Behaviour {
+    shared: Arc<Mutex<Shared>>,
+    receiver: mpsc::Receiver<PeerId>,
+    supported_protocols: HashMap<StreamProtocol, mpsc::Sender<(PeerId, Stream)>>,
+
+    events: VecDeque<ToSwarm<(), RegisterProtocol>>,
+}
+
+pub(crate) struct Shared {
+    connections: HashMap<ConnectionId, PeerId>,
+    senders: HashMap<ConnectionId, mpsc::Sender<NewStream>>,
+
+    pending_channels: HashMap<PeerId, (mpsc::Sender<NewStream>, mpsc::Receiver<NewStream>)>,
+
+    dial_sender: mpsc::Sender<PeerId>,
+}
+
+impl Shared {
+    pub(crate) fn new(dial_sender: mpsc::Sender<PeerId>) -> Self {
+        Self {
+            connections: Default::default(),
+            senders: Default::default(),
+            pending_channels: Default::default(),
+            dial_sender,
+        }
+    }
+
+    pub(crate) fn sender(&mut self, peer: PeerId) -> mpsc::Sender<NewStream> {
+        let maybe_sender = self
+            .connections
+            .iter()
+            .filter_map(|(c, p)| (p == &peer).then_some(c))
+            .choose(&mut rand::thread_rng())
+            .and_then(|c| self.senders.get(c));
+
+        match maybe_sender {
+            Some(sender) => {
+                tracing::debug!("Returning sender to existing connection");
+
+                sender.clone()
+            }
+            None => {
+                let (sender, _) = self
+                    .pending_channels
+                    .entry(peer)
+                    .or_insert_with(|| mpsc::channel(0));
+
+                let _ = self.dial_sender.try_send(peer);
+
+                sender.clone()
+            }
+        }
+    }
+
+    fn receiver(&mut self, peer: PeerId, connection: ConnectionId) -> mpsc::Receiver<NewStream> {
+        if let Some((sender, receiver)) = self.pending_channels.remove(&peer) {
+            tracing::debug!(%peer, %connection, "Returning existing pending receiver");
+
+            self.senders.insert(connection, sender);
+            return receiver;
+        }
+
+        tracing::debug!(%peer, %connection, "Creating new channel pair");
+
+        let (sender, receiver) = mpsc::channel(0);
+        self.senders.insert(connection, sender);
+
+        receiver
+    }
+}
+
+impl Behaviour {
+    pub fn new() -> Self {
+        let (sender, receiver) = mpsc::channel(0);
+
+        Self {
+            shared: Arc::new(Mutex::new(Shared::new(sender))),
+            receiver,
+            supported_protocols: Default::default(),
+            events: Default::default(),
+        }
+    }
+
+    /// Obtain a new [`Control`] for the provided protocol.
+    ///
+    /// A [`Control`] only deals with the _outbound_ side of a protocol.
+    /// To accept inbound streams for a protocol, use [`Behaviour::accept`].
+    pub fn new_control(&self, protocol: StreamProtocol) -> Control {
```

Comment:
I think I want to move the protocol argument to `Control::open_stream`. It seems unnecessarily limiting to force the user to create multiple `Control`s.